README.md
# The von Koch snowflake

This project generates a SVG file with a von Koch snowflake, using the [cairo-fortran](https://github.com/vmagnin/cairo-fortran) bindings as a fpm (Fortran Package Manager) dependency.

The classical recursive algorithm is used, each line segment being replaced by four segments.

![A von Koch snowflake](figures/von_koch_snowflake.svg)

Be careful with the number of recursions: the SVG file is growing like $3 \times 4^n$ ! Six recursions will generate a 2.4 Mio SVG file. Your SVG viewer may quickly freeze or crash...

## Requirements and dependencies

You need:

* a modern Fortran compiler, for example GFortran or the Intel ifort/ifx compilers. See the [Fortran-lang.org compilers page](https://fortran-lang.org/compilers/) for other compilers.
* The Cairo development files (`libcairo2-dev` package in Ubuntu).
* The Fortran Package Manager [fpm](https://fpm.fortran-lang.org/).

## Running the program

Just type `fpm run` and fpm will manage the dependencies and build and run the program:

```bash
$ fpm run
 + mkdir -p build/dependencies
Initialized empty Git repository in von_koch_snowflake/build/dependencies/cairo-fortran/.git/
remote: Enumerating objects: 23, done.
remote: Counting objects: 100% (23/23), done.
remote: Compressing objects: 100% (22/22), done.
remote: Total 23 (delta 0), reused 14 (delta 0), pack-reused 0
Unpacking objects: 100% (23/23), 82.99 KiB | 2.37 MiB/s, done.
From https://github.com/vmagnin/cairo-fortran
 * branch            HEAD       -> FETCH_HEAD
cairo-enums.f90                        done.
cairo-auto.f90                         done.
von_koch.f90                           done.
libvon_koch_snowflake.a                done.
main.f90                               done.
von_koch_snowflake                     done.
[100%] Project compiled successfully.
 ----------------------------------
           6 recursions
 ----------------------------------
 Output file: von_koch_snowflake.svg
 ----------------------------------
```


## Licenses

This project is licensed under the [GNU General Public License version 3 or later](http://www.gnu.org/licenses/gpl.html).

The documentation is under the [GNU Free Documentation License 1.3 or any later version](http://www.gnu.org/licenses/fdl.html). The figures are under [CC-BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) license.

## Cairo documentation

* https://cairographics.org/documentation/
* API Reference Manual: https://cairographics.org/manual/

## Bibliography

* https://en.wikipedia.org/wiki/Koch_snowflake

CHANGELOG.md
# Changelog
All notable changes to the gtk-fortran project are documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [von Koch snowflake 1.0] 2023-06-06

* Fortran version, using cairo-fortran.


src/von_koch.f90
! This file is part of the von_koch_snowflake Fortran project.
! Copyright (C) 2023 Vincent MAGNIN
!
! This is free software; you can redistribute it and/or modify
! it under the terms of the GNU General Public License as published by
! the Free Software Foundation; either version 3, or (at your option)
! any later version.
!
! This software is distributed in the hope that it will be useful,
! but WITHOUT ANY WARRANTY; without even the implied warranty of
! MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
! GNU General Public License for more details.
!
! You should have received a copy of the GNU General Public License along with
! this program; see the files LICENSE and LICENSE_EXCEPTION respectively.
! If not, see <http://www.gnu.org/licenses/>.
!------------------------------------------------------------------------------
! Contributed by Vincent Magnin
! Pascal version: 1987 ?
! Python version: 2010
! Fortran version: 2023-06-05
! Last modifications: 2023-06-06
!------------------------------------------------------------------------------

module von_koch
    ! Cairo uses C doubles:
    use, intrinsic :: iso_c_binding, only: dp=>c_double, c_ptr
    ! Use the cairo-fortran bindings:
    use cairo

    implicit none
    private
    public :: von_koch_segment, ROTATE

    ! The y axis going downward in Cairo, the 60° rotation must be inverted:
    complex(dp), parameter :: ROTATE = cmplx(0.5_dp, -sqrt(3._dp) / 2, dp)

contains

    subroutine draw_segment(P1, P2, r, g, b, cr)
        complex(dp), intent(in) :: P1, P2
        real(dp), intent(in)    :: r, g, b
        ! Cairo context:
        type(c_ptr), intent(in) :: cr

        call cairo_set_source_rgb(cr, r, g, b)
        call cairo_move_to(cr, P1%re, P1%im)
        call cairo_line_to(cr, P2%re, P2%im)
        call cairo_stroke(cr)
    end subroutine draw_segment


    recursive subroutine von_koch_segment(E, F, n, nmax, cr)
        complex(dp), intent(in) :: E, F
        integer, intent(in)     :: n, nmax
        type(c_ptr), intent(in) :: cr
        complex(dp) :: A, B, C

        ! F---A---C---E
        !      \ /
        !       B

        if (n < nmax) then      ! Recursion case
            C = (F + 2*E) / 3
            A = (2*F + E) / 3
            B = A + (C - A) * ROTATE
            call von_koch_segment(E, C, n+1, nmax, cr)
            call von_koch_segment(C, B, n+1, nmax, cr)
            call von_koch_segment(B, A, n+1, nmax, cr)
            call von_koch_segment(A, F, n+1, nmax, cr)
        else                    ! Terminating case
            call draw_segment(E, F, 0._dp, 0._dp, 0._dp, cr)
        end if
    end subroutine

end module von_koch


other_languages/vonkoch.py
# This file is part of the von_koch_snowflake Fortran project.
# Copyright (C) 2023 Vincent MAGNIN
#
# This is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3, or (at your option)
# any later version.
#
# This software is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# this program; see the files LICENSE and LICENSE_EXCEPTION respectively.
# If not, see <http://www.gnu.org/licenses/>.
#------------------------------------------------------------------------------
# Contributed by Vincent Magnin
# Pascal version: 1987 ?
# Python version: 2010
# Fortran version: 2023-06-05
# Last modifications: 2023-06-06
#------------------------------------------------------------------------------

def koch(l,n):
    # Fractale de Koch
    if n<=0:
        forward(l)
    else:
        ll=l/3
        nn=n-1
        koch(ll,nn)
        left(60)
        koch(ll,nn)
        right(120)
        koch(ll,nn)
        left(60)
        koch(ll,nn)

def flocon(l,n):
    # Flocon de Koch
    koch(l,n)
    right(120)
    koch(l,n)
    right(120)
    koch(l,n)

# programme principal
from turtle import *
speed(0)
hideturtle()

etape=int(input("Entrez l'ordre de l'étape du flocon de Von Koch "))
taille1=float(input("Entrez la taille du côté du triangle initial "))
flocon(taille1,etape)

# on utilise une méthode récursive pour programmer les fonctions

print("Appuyez sur Entrée")
pause=input()   # Pour éviter que la fenêtre graphique ne se ferme à la fin du programme !


other_languages/VONKOCH.PAS
// This file is part of the von_koch_snowflake Fortran project.
// Copyright (C) 2023 Vincent MAGNIN
//
// This is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3, or (at your option)
// any later version.
//
// This software is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along with
// this program; see the files LICENSE and LICENSE_EXCEPTION respectively.
// If not, see <http://www.gnu.org/licenses/>.
//------------------------------------------------------------------------------
// Contributed by Vincent Magnin
// Pascal version: 1987 ?
// Python version: 2010
// Fortran version: 2023-06-05
// Last modifications: 2023-06-06
//------------------------------------------------------------------------------

{$i c:graph.p};
var
   xc,yc,xa,ya,xb,yb:integer;

procedure vonkoch(xe,ye,xf,yf:integer;n:integer);
var xa,ya,xb,yb,xc,yc:integer;
begin
     if n<4 then
     begin
          n:=n+1;
          xc:=round(1/3*(xf+2*xe));
          yc:=round(1/3*(yf+2*ye));
          xa:=round(1/3*(2*xf+xe));
          ya:=round(1/3*(2*yf+ye));
          xb:=round(1/2*(xf+xe)-1/(2*sqrt(3))*(yf-ye));
          yb:=round(1/2*(yf+ye)+1/(2*sqrt(3))*(xf-xe));
          vonkoch(xe,ye,xc,yc,n);
          vonkoch(xc,yc,xb,yb,n);
          vonkoch(xb,yb,xa,ya,n);
          vonkoch(xa,ya,xf,yf,n);
     end
     else
     begin
          draw(round(xe*1/0.43),ye,round(xf*1/0.43),yf,1);
     end;
end;

begin
     hires;
     xc:=100;
     yc:=50;
     xa:=250;
     ya:=50;
     xb:=round(1/2*(xa-xc)-sqrt(3)/2*(ya-yc)+xc);
     yb:=round(1/2*(ya-yc)+sqrt(3)/2*(xa-xc)+yc);
     vonkoch(xa,ya,xc,yc,0);
     vonkoch(xb,yb,xa,ya,0);
     vonkoch(xc,yc,xb,yb,0);
     fillshape(300,100,1,1);
     repeat
     until keypressed;

end.


app/main.f90
! This file is part of the von_koch_snowflake Fortran project.
! Copyright (C) 2023 Vincent MAGNIN
!
! This is free software; you can redistribute it and/or modify
! it under the terms of the GNU General Public License as published by
! the Free Software Foundation; either version 3, or (at your option)
! any later version.
!
! This software is distributed in the hope that it will be useful,
! but WITHOUT ANY WARRANTY; without even the implied warranty of
! MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
! GNU General Public License for more details.
!
! You should have received a copy of the GNU General Public License along with
! this program; see the files LICENSE and LICENSE_EXCEPTION respectively.
! If not, see <http://www.gnu.org/licenses/>.
!------------------------------------------------------------------------------
! Contributed by Vincent Magnin
! Pascal version: 1987 ?
! Python version: 2010
! Fortran version: 2023-06-05
! Last modifications: 2023-06-06
!------------------------------------------------------------------------------

program main
    ! Use the cairo-fortran bindings:
    use cairo
    use cairo_enums
    ! Cairo uses C doubles:
    use, intrinsic :: iso_c_binding, only: dp=>c_double, c_ptr
    use von_koch, only: von_koch_segment, ROTATE

    implicit none
    ! Cairo surface and Cairo context:
    type(c_ptr) :: surface, cr
    character(*), parameter :: FILENAME = "von_koch_snowflake.svg"
    ! Size of the surface in mm (landscape A4 paper):
    real(dp), parameter :: IMAGE_WIDTH  = 210._dp
    real(dp), parameter :: IMAGE_HEIGHT = 297._dp

    ! Be very careful, 6 recursions will generate a 2.4 Mio SVG file,
    ! and it grows like 3*4^n...
    integer, parameter  :: MAX_RECURSION = 6
    ! The triangle vertices:
    complex(dp) :: A, B, C

    ! The object will be rendered in a SVG file:
    surface = cairo_svg_surface_create(FILENAME//c_null_char, &
                                      & IMAGE_WIDTH, IMAGE_HEIGHT)
    call cairo_svg_surface_set_document_unit(surface, CAIRO_SVG_UNIT_MM)
    cr = cairo_create(surface)

    ! Initialize parameters:
    call cairo_set_antialias(cr, CAIRO_ANTIALIAS_BEST)
    call cairo_set_line_width(cr, 1._dp / MAX_RECURSION**2)

    ! Remarks: 
    !   * The y axis is oriented downward (and rotation angles are inverted).
    !   * We work in the complex plane.
    !
    ! We start with an equilateral triangle:
    ! C---A
    !  \ /
    !   B

    ! We need margins for printing:
    C = cmplx(IMAGE_WIDTH/10,       IMAGE_HEIGHT/5, dp)
    A = cmplx(IMAGE_WIDTH*9._dp/10, IMAGE_HEIGHT/5, dp)
    B = A + (C - A) * ROTATE

    ! We launch the recursive algorithm on each side of the triangle:
    call von_koch_segment(A, C, 0, MAX_RECURSION, cr)
    call von_koch_segment(B, A, 0, MAX_RECURSION, cr)
    call von_koch_segment(C, B, 0, MAX_RECURSION, cr)

    ! Finalyze:
    call cairo_destroy(cr)
    call cairo_surface_destroy(surface)

    print *, "----------------------------------"
    print *, MAX_RECURSION, "recursions"
    print *, "----------------------------------"
    print *, "Output file: ", FILENAME
    print *, "----------------------------------"
end program main
